tests/unit/s2n_x509_validator_test.c

```diff
@@ -1407,6 +1412,96 @@ int main(int argc, char **argv) {
         /* Expect no crash. */
     }
 
+    /* Test validator trusts a SHA-1 root signature with no SHA-1 signatures on the other certificates */
+    {
+        struct s2n_x509_trust_store trust_store;
+        s2n_x509_trust_store_init_empty(&trust_store);
+        EXPECT_EQUAL(0, s2n_x509_trust_store_from_ca_file(&trust_store, S2N_SHA1_ROOT_SIGNATURE_CERT_CHAIN, NULL));
+
+        struct s2n_x509_validator validator;
+        s2n_x509_validator_init(&validator, &trust_store, 1);
+
+        struct s2n_config *config = s2n_config_new();
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+
+        struct s2n_connection *connection = s2n_connection_new(S2N_CLIENT);
+        EXPECT_SUCCESS(s2n_connection_set_config(connection, config));
+
+        EXPECT_NOT_NULL(connection);
+        connection->actual_protocol_version = S2N_TLS13;
+
+        struct host_verify_data verify_data = { .callback_invoked = 0, .found_name = 0, .name = NULL };
+        EXPECT_SUCCESS(s2n_connection_set_verify_host_callback(connection, verify_host_accept_everything, &verify_data));
+        
+        uint8_t cert_chain_pem[S2N_MAX_TEST_PEM_SIZE];
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_SHA1_ROOT_SIGNATURE_CERT_CHAIN, (char *) cert_chain_pem, S2N_MAX_TEST_PEM_SIZE));
+        struct s2n_stuffer chain_stuffer;
+        uint32_t chain_len = write_pem_file_to_stuffer_as_chain(&chain_stuffer, (const char *) cert_chain_pem, S2N_TLS13);
+        EXPECT_TRUE(chain_len > 0);
+        uint8_t *chain_data = s2n_stuffer_raw_read(&chain_stuffer, (uint32_t) chain_len);
+
+        struct s2n_pkey public_key_out;
+        EXPECT_SUCCESS(s2n_pkey_zero_init(&public_key_out));
+        s2n_pkey_type pkey_type;
+        EXPECT_EQUAL(S2N_CERT_OK,
+                     s2n_x509_validator_validate_cert_chain(&validator, connection, chain_data, chain_len, &pkey_type, &public_key_out));
+        EXPECT_EQUAL(S2N_CERT_OK,
+                     s2n_x509_validator_validate_certificate_signatures(&validator, connection));
```

Comment:
Do you call s2n_x509_validator_validate_cert_chain before s2n_x509_validator_validate_certificate_signatures because you HAVE to call s2n_x509_validator_validate_cert_chain in order to ensure the certificates are properly ordered?

If yes, then s2n_x509_validator_validate_certificate_signatures should be called automatically from s2n_x509_validator_validate_cert_chain so that the caller doesn't have to know to call s2n_x509_validator_validate_cert_chain first.

If no, then don't call s2n_x509_validator_validate_cert_chain here.